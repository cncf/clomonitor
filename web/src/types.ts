export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
