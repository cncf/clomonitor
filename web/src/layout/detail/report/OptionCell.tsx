import { isBoolean, isNull, isUndefined } from 'lodash';
import { FaRegCheckCircle, FaRegQuestionCircle, FaRegTimesCircle } from 'react-icons/fa';

import { REPORT_OPTIONS } from '../../../data';
import { ReportOption, ReportOptionData } from '../../../types';
import ElementWithTooltip from '../../common/ElementWithTooltip';
import styles from './OptionCell.module.css';

interface Props {
  label: ReportOption;
  value: boolean | string;
}

function getOptionInfo(key: ReportOption) {
  return REPORT_OPTIONS[key];
}

const OptionCell = (props: Props) => {
  const errorIcon = <FaRegTimesCircle data-testid="error-icon" className={`text-danger ${styles.icon}`} />;
  const successIcon = <FaRegCheckCircle data-testid="success-icon" className={`text-success ${styles.icon}`} />;

  const opt: ReportOptionData = getOptionInfo(props.label);

  const checkStatus = (): boolean => {
    if (isBoolean(props.value)) {
      if (props.value) {
        return true;
      } else {
        return false;
      }
    } else {
      if (isUndefined(props.value) || isNull(props.value) || props.value === '') {
        return false;
      } else {
        return true;
      }
    }
  };

  const status: boolean = checkStatus();

  return (
    <tr>
      <td className={`text-center ${styles.iconCell}`}>{status ? successIcon : errorIcon}</td>
      <td>
        <div className={`d-table w-100 ${styles.contentCell}`}>
          <div className="d-flex flex-row align-items-baseline align-items-lg-center">
            <div className="text-muted me-2">{opt.icon}</div>
            <div className="d-flex flex-column align-items-start flex-grow-1 truncateWrapper">
              <div data-testid="opt-name" className={`d-flex flex-row align-items-center mt-1 ${styles.name}`}>
                {(() => {
                  switch (props.label) {
                    case ReportOption.SPDX:
                      return <small className="fw-bold">{props.value || 'Not detected'}</small>;

                    default:
                      return <small className="fw-bold">{opt.name}</small>;
                  }
                })()}
              </div>
              <div className={`d-none d-lg-block text-muted text-truncate mt-1 w-100 ${styles.legend}`}>
                {opt.legend}
              </div>
            </div>
          </div>
        </div>
      </td>
      <td className={`text-center text-muted ${styles.iconCell}`}>
        <ElementWithTooltip
          className="ms-2 lh-1"
          element={<FaRegQuestionCircle className={styles.icon} />}
          tooltipWidth={285}
          tooltipClassName={styles.tooltipMessage}
          tooltipMessage={<div className="my-2">{opt.description}</div>}
          visibleTooltip
          active
        />
      </td>
    </tr>
  );
};

export default OptionCell;
