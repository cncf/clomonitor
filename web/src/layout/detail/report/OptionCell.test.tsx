import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ReportOption } from '../../../types';
import OptionCell from './OptionCell';
jest.mock('react-markdown', () => () => <div>markdown</div>);
jest.mock('rehype-external-links', () => () => <></>);

const defaultProps = {
  label: ReportOption.Adopters,
  check: {
    passed: true,
  },
};

const user = userEvent.setup({ delay: null });

describe('OptionCell', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <table>
        <tbody>
          <OptionCell {...defaultProps} />
        </tbody>
      </table>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders option', () => {
      render(
        <table>
          <tbody>
            <OptionCell {...defaultProps} />
          </tbody>
        </table>
      );

      expect(screen.getByText('Adopters')).toBeInTheDocument();
      expect(screen.getByTestId('opt-name')).toBeInTheDocument();
      expect(
        screen.getByText('List of organizations using this project in production or at stages of testing')
      ).toBeInTheDocument();
    });

    it('renders special SPDX option', () => {
      render(
        <table>
          <tbody>
            <OptionCell
              label={ReportOption.SPDX}
              check={{
                passed: true,
                value: 'Apache-2.0',
              }}
            />
          </tbody>
        </table>
      );

      expect(screen.getByText('Apache-2.0')).toBeInTheDocument();
      expect(screen.getByText(/file contains the repository's license/)).toBeInTheDocument();
      expect(screen.getByText('LICENSE')).toBeInTheDocument();
    });

    it('renders option with url', () => {
      render(
        <table>
          <tbody>
            <OptionCell
              {...defaultProps}
              check={{
                url: 'https://github.com/project-akri/akri/blob/main/ADOPTERS.md',
                exempt: false,
                failed: false,
                passed: true,
              }}
            />
          </tbody>
        </table>
      );

      const link = screen.getByRole('link', { name: 'Checks reference documentation' });
      expect(link).toBeInTheDocument();
      expect(link).toHaveTextContent('Adopters');
      expect(link).toHaveProperty('target', '_blank');
      expect(link).toHaveProperty('href', 'https://github.com/project-akri/akri/blob/main/ADOPTERS.md');
      expect(link).toHaveProperty('rel', 'noopener noreferrer');
    });

    it('renders option with details', async () => {
      jest.useFakeTimers();

      render(
        <table>
          <tbody>
            <OptionCell
              label={ReportOption.DependencyUpdateTool}
              check={{
                exempt: false,
                failed: false,
                passed: false,
                details:
                  '### Determines if the project uses a dependency update tool\n\n**OpenSSF Scorecard score**: 0\n**Reason**: no update tool detected\n\n**Details**:\n\nWarn: dependabot config file not detected in source location.\n\t\t\tWe recommend setting this configuration in code so it can be easily verified by others.\nWarn: renovatebot config file not detected in source location.\n\t\t\tWe recommend setting this configuration in code so it can be easily verified by others.\n\n*Please see the [check docs](https://github.com/ossf/scorecard/blob/33f80c93dc79f860d874857c511c4d26d399609d/docs/checks.md#dependency-update-tool) for more details*',
              }}
            />
          </tbody>
        </table>
      );

      const icons = screen.getAllByTestId('error-icon');
      expect(icons).toHaveLength(2);

      const dropdown = screen.getByRole('complementary');

      expect(dropdown).not.toHaveClass('show');

      await user.hover(icons[0]);

      act(() => {
        jest.advanceTimersByTime(100);
      });

      expect(dropdown).toHaveClass('show');
      expect(screen.getByText('markdown')).toBeInTheDocument();

      jest.useRealTimers();
    });

    describe('passed', () => {
      it('when true: renders option with success icon', () => {
        render(
          <table>
            <tbody>
              <OptionCell {...defaultProps} />
            </tbody>
          </table>
        );

        expect(screen.getByTestId('success-icon')).toBeInTheDocument();
        expect(screen.queryByTestId('error-icon')).toBeNull();
      });

      it('when false: renders option with error icon', () => {
        render(
          <table>
            <tbody>
              <OptionCell
                {...defaultProps}
                check={{
                  passed: false,
                }}
              />
            </tbody>
          </table>
        );

        expect(screen.getByTestId('error-icon')).toBeInTheDocument();
        expect(screen.queryByTestId('success-icon')).toBeNull();
      });
    });

    describe('exempt', () => {
      it('when true', () => {
        render(
          <table>
            <tbody>
              <OptionCell
                label={ReportOption.Adopters}
                check={{
                  passed: false,
                  exempt: true,
                  exemption_reason: 'this is a sample reason',
                }}
              />
            </tbody>
          </table>
        );

        expect(screen.getAllByTestId('exempt-icon')).toHaveLength(2);
        expect(screen.queryByTestId('failed-icon')).toBeNull();
        expect(screen.queryByTestId('success-icon')).toBeNull();
        expect(screen.queryByTestId('error-icon')).toBeNull();
      });

      it('displays reason tooltip', async () => {
        jest.useFakeTimers();

        render(
          <table>
            <tbody>
              <OptionCell
                label={ReportOption.Adopters}
                check={{
                  passed: false,
                  exempt: true,
                  exemption_reason: 'this is a sample reason',
                }}
              />
            </tbody>
          </table>
        );

        const icon = screen.getByTestId('elementWithTooltip');
        await user.hover(icon);

        expect(await screen.findByRole('tooltip')).toBeInTheDocument();
        expect(screen.getByText('This repository is exempt from passing this check')).toBeInTheDocument();
        expect(screen.getByText('Reason:')).toBeInTheDocument();
        expect(screen.getByText(/this is a sample reason/)).toBeInTheDocument();

        jest.useRealTimers();
      });
    });

    describe('failed', () => {
      it('when true', () => {
        render(
          <table>
            <tbody>
              <OptionCell
                label={ReportOption.Adopters}
                check={{
                  passed: false,
                  failed: true,
                  fail_reason: 'this is a sample reason',
                }}
              />
            </tbody>
          </table>
        );

        expect(screen.getAllByTestId('failed-icon')).toHaveLength(2);
        expect(screen.queryByTestId('exempt-icon')).toBeNull();
        expect(screen.queryByTestId('success-icon')).toBeNull();
        expect(screen.queryByTestId('error-icon')).toBeNull();
      });

      it('displays reason tooltip', async () => {
        jest.useFakeTimers();

        render(
          <table>
            <tbody>
              <OptionCell
                label={ReportOption.Adopters}
                check={{
                  passed: false,
                  failed: true,
                  fail_reason: 'this is a sample reason',
                }}
              />
            </tbody>
          </table>
        );

        const icon = screen.getByTestId('elementWithTooltip');
        await user.hover(icon);

        expect(await screen.findByRole('tooltip')).toBeInTheDocument();
        expect(screen.getByText('Something went wrong running this check')).toBeInTheDocument();
        expect(screen.getByText('Reason:')).toBeInTheDocument();
        expect(screen.getByText(/this is a sample reason/)).toBeInTheDocument();

        jest.useRealTimers();
      });
    });
  });
});
