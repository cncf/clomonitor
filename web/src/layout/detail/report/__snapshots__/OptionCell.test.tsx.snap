// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionCell creates snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="text-center iconCell"
        >
           
          <svg
            class="text-success icon"
            data-testid="success-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 48c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m140.204 130.267l-22.536-22.718c-4.667-4.705-12.265-4.736-16.97-.068L215.346 303.697l-59.792-60.277c-4.667-4.705-12.265-4.736-16.97-.069l-22.719 22.536c-4.705 4.667-4.736 12.265-.068 16.971l90.781 91.516c4.667 4.705 12.265 4.736 16.97.068l172.589-171.204c4.704-4.668 4.734-12.266.067-16.971z"
            />
          </svg>
        </td>
        <td
          class="pe-4"
        >
          <div
            class="d-table w-100 contentCell"
          >
            <div
              class="d-flex flex-row align-items-baseline align-items-md-center"
            >
              <div
                class="text-muted me-2"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  version="1.1"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 16h8v-16h-8v16zM5 2h2v2h-2v-2zM5 6h2v2h-2v-2zM5 10h2v2h-2v-2zM1 2h2v2h-2v-2zM1 6h2v2h-2v-2zM1 10h2v2h-2v-2zM9 5h7v1h-7zM9 16h2v-4h3v4h2v-9h-7z"
                  />
                </svg>
              </div>
              <div
                class="d-flex flex-column align-items-start flex-grow-1 truncateWrapper"
              >
                <div
                  class="d-flex flex-row align-items-center w-100 name"
                  data-testid="opt-name"
                >
                  <small
                    class="fw-bold text-truncate"
                  >
                    Adopters
                  </small>
                </div>
                <div
                  class="d-none d-md-flex flex-row text-muted w-100 legend"
                >
                  <div
                    class="text-truncate"
                  >
                    <span>
                      List of organizations using this project in production or at stages of testing
                    </span>
                  </div>
                  <div
                    class="d-none d-lg-flex text-nowrap"
                  >
                    <a
                      aria-label="Open external link"
                      class="link d-inline w-100 ms-1"
                      href="/docs/topics/checks/#adopters"
                      rel="noopener noreferrer"
                      tabindex="-1"
                      target="_blank"
                    >
                      <div
                        class="d-flex flex-row align-items-baseline"
                      >
                        <div
                          class="d-flex flex-row align-items-center w-100"
                        >
                          <div>
                            [check docs
                          </div>
                          <svg
                            class="ms-1 extraMiniIcon"
                            fill="none"
                            height="1em"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                            viewBox="0 0 24 24"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                            />
                            <polyline
                              points="15 3 21 3 21 9"
                            />
                            <line
                              x1="10"
                              x2="21"
                              y1="14"
                              y2="3"
                            />
                          </svg>
                          <div>
                            ]
                          </div>
                        </div>
                      </div>
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
